File: archiso/airootfs/usr/share/churros/preferences/pages/backup.py
import sys
import os
import threading

ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), ".."))
PREF = os.path.abspath(os.path.join(ROOT, "preferences"))

sys.path.insert(0, ROOT)
sys.path.insert(0, PREF)

import gi

gi.require_version("Gtk", "4.0")

from gi.repository import Gtk, GLib, Gio

from widgets.page import Page
from widgets.group import Group
from widgets.row import Row

from services.backup_service import BackupService


class BackupPage(Page):

    def __init__(self, navigator):

        super().__init__(
            navigator,
            "Copia de seguridad",
            "Exporta, importa o restablece la configuracion de ChurrOS",
            parent_page="system"
        )

        self._build()

    def _build(self):

        #
        # Exportar
        #

        export_group = Group("Exportar")

        export_row = Row(
            title="Exportar configuracion",
            subtitle="Empaqueta ajustes y dotfiles en un archivo .tar",
            icon="backup.svg",
            callback=lambda *_: self._on_export()
        )

        export_group.add(export_row)

        self.add(export_group)

        #
        # Importar
        #

        import_group = Group("Importar")

        import_row = Row(
            title="Importar configuracion",
            subtitle="Restaurar desde un backup de ChurrOS (.tar)",
            icon="backup.svg",
            callback=lambda *_: self._on_import()
        )

        import_group.add(import_row)

        self.add(import_group)

        #
        # Restablecer
        #

        reset_group = Group("Restablecer")

        reset_row = Row(
            title="Restablecer a valores de fabrica",
            subtitle="Borra tus cambios y restaura los defaults de ChurrOS",
            icon="backup.svg",
            callback=lambda *_: self._on_reset()
        )

        reset_group.add(reset_row)

        self.add(reset_group)

        #
        # Estado
        #

        status_group = Group("Estado")

        self._status_label = Gtk.Label(
            label="Listo."
        )
        self._status_label.set_xalign(0)
        self._status_label.set_wrap(True)
        self._status_label.add_css_class("row-subtitle")
        self._status_label.set_margin_start(14)
        self._status_label.set_margin_end(14)
        self._status_label.set_margin_top(10)
        self._status_label.set_margin_bottom(10)

        status_group.add(self._status_label)

        self.add(status_group)

    def _set_status(self, msg):

        self._status_label.set_label(msg)

    def _on_export(self):

        dialog = Gtk.FileDialog()
        dialog.set_title("Guardar backup")

        filter_tar = Gtk.FileFilter()
        filter_tar.set_name("Archivo tar")
        filter_tar.add_pattern("*.tar")
        filter_tar.add_pattern("*.tar.gz")
        filter_tar.add_pattern("*.tar.zst")

        filters = Gio.ListStore.new(Gtk.FileFilter)
        filters.append(filter_tar)
        dialog.set_filters(filters)
        dialog.set_default_filter(filter_tar)

        try:

            dialog.set_initial_name(
                "churros-backup.tar"
            )

            dialog.set_initial_folder(
                Gio.File.new_for_path(
                    os.path.expanduser("~")
                )
            )

        except Exception:
            pass

        def on_result(source, result, _user_data=None):

            try:

                file = dialog.save_finish(result)

            except GLib.Error:
                return

            if file is None:
                return

            path = file.get_path() if hasattr(file, "get_path") else None

            if not path:
                return

            self._set_status("Exportando...")

            def worker():

                try:

                    BackupService.export_to(path)

                    GLib.idle_add(
                        lambda: self._set_status(
                            "Backup guardado en " + path
                        )
                    )

                except Exception as exc:

                    GLib.idle_add(
                        lambda: self._set_status(
                            "Error al exportar: " + str(exc)
                        )
                    )


            threading.Thread(
                target=worker,
                daemon=True
            ).start()

        dialog.save(self.get_root(), None, on_result)

    def _on_import(self):

        dialog = Gtk.FileDialog()
        dialog.set_title("Seleccionar backup de ChurrOS")

        filter_tar = Gtk.FileFilter()
        filter_tar.set_name("Archivo tar")
        filter_tar.add_pattern("*.tar")
        filter_tar.add_pattern("*.tar.gz")
        filter_tar.add_pattern("*.tar.zst")

        filters = Gio.ListStore.new(Gtk.FileFilter)
        filters.append(filter_tar)
        dialog.set_filters(filters)
        dialog.set_default_filter(filter_tar)

        try:

            dialog.set_initial_folder(
                Gio.File.new_for_path(
                    os.path.expanduser("~")
                )
            )

        except Exception:
            pass

        def on_result(source, result, _user_data=None):

            try:

                file = dialog.open_finish(result)

            except GLib.Error:
                return

            if file is None:
                return

            path = file.get_path() if hasattr(file, "get_path") else None

            if not path or not os.path.isfile(path):
                return

            self._set_status("Importando... recargando tras aplicar")

            confirm = Gtk.AlertDialog()
            confirm.set_heading("Importar configuracion")
            confirm.set_message(
                "Esto reemplazara tu configuracion actual "
                "con la del archivo. ¿Continuar?"
            )
            confirm.set_modal(True)
            confirm.set_buttons(["Cancelar", "Importar"])

            def on_confirm(d, r):

                try:
                    response = d.choose_finish(r)
                except Exception:
                    return

                if response != 1:
                    return

                def worker():

                    try:

                        BackupService.import_from(path)

                        GLib.idle_add(
                            lambda: self._set_status(
                                "Configuracion importada."
                                " Reinicia las apps para ver todos"
                                " los cambios."
                            )
                        )

                    except Exception as exc:

                        GLib.idle_add(
                            lambda: self._set_status(
                                "Error al importar: " + str(exc)
                            )
                        )

                threading.Thread(
                    target=worker,
                    daemon=True
                ).start()

            confirm.choose(self.get_root(), None, on_confirm)

        dialog.open(self.get_root(), None, on_result)

    def _on_reset(self):

        dialog = Gtk.AlertDialog()
        dialog.set_heading("Restablecer a valores de fabrica")
        dialog.set_message(
            "Se borraran tus ajustes personales (tema, wallpaper, "
            "tipografia, dotfiles de niri/foot/fuzzel/mako/waybar) "
            "y se restauraran los defaults de ChurrOS. ¿Continuar?"
        )
        dialog.set_modal(True)
        dialog.set_buttons(["Cancelar", "Restablecer"])

        def on_response(d, result):

            try:
                response = d.choose_finish(result)
            except Exception:
                return

            if response != 1:
                return

            self._set_status("Restableciendo...")

            def worker():

                try:

                    BackupService.reset_to_defaults()

                    GLib.idle_add(
                        lambda: self._set_status(
                            "Configuraciones restablecidas a defaults. "
                            "Reinicia las apps para ver todos los cambios."
                        )
                    )

                except Exception as exc:

                    GLib.idle_add(
                        lambda: self._set_status(
                            "Error al restablecer: " + str(exc)
                        )
                    )


            threading.Thread(
                target=worker,
                daemon=True
            ).start()

        dialog.choose(self.get_root(), None, on_response)

File: rust/preferences/src/pages/backup.rs
// ==========================================
// BackupPage — exportar/importar/restablecer configuracion
// (equivalente a pages/backup.py)
// ==========================================

use gtk::prelude::*;

use std::cell::RefCell;
use std::path::PathBuf;

use crate::services::backup_service::BackupService;
use crate::widgets::group::Group;
use crate::widgets::page::Page;
use crate::widgets::row::Row;

// Equivalente a self._status_label (los callbacks son Fn(&gtk::Button),
// asi que el label se comparte via thread_local).
thread_local! {
    static STATUS_LABEL: RefCell<Option<gtk::Label>> = RefCell::new(None);
}

struct WLabel(gtk::Label);
impl crate::widgets::AsWidget for WLabel {
    fn widget(&self) -> &gtk::Widget {
        self.0.upcast_ref()
    }
}

pub fn build(navigator: gtk::Stack) -> Page {
    let page = Page::new(
        Some(navigator),
        "Copia de seguridad",
        Some("Exporta, importa o restablece la configuracion de ChurrOS"),
        Some("system".to_string()),
    );

    // ---------- Exportar ----------
    let mut export_group = Group::new("Exportar");

    export_group.add(&Row::new(
        "Exportar configuracion",
        Some("Empaqueta ajustes y dotfiles en un archivo .tar"),
        Some("backup.svg"),
        None,
        None,
        Some(Box::new(on_export)),
    ));

    page.add(export_group.widget());

    // ---------- Importar ----------
    let mut import_group = Group::new("Importar");

    import_group.add(&Row::new(
        "Importar configuracion",
        Some("Restaurar desde un backup de ChurrOS (.tar)"),
        Some("backup.svg"),
        None,
        None,
        Some(Box::new(on_import)),
    ));

    page.add(import_group.widget());

    // ---------- Restablecer ----------
    let mut reset_group = Group::new("Restablecer");

    reset_group.add(&Row::new(
        "Restablecer a valores de fabrica",
        Some("Borra tus cambios y restaura los defaults de ChurrOS"),
        Some("backup.svg"),
        None,
        None,
        Some(Box::new(on_reset)),
    ));

    page.add(reset_group.widget());

    // ---------- Estado ----------
    let mut status_group = Group::new("Estado");

    let status_label = gtk::Label::new(Some("Listo."));
    status_label.set_xalign(0.0);
    status_label.set_wrap(true);
    status_label.add_css_class("row-subtitle");
    status_label.set_margin_start(14);
    status_label.set_margin_end(14);
    status_label.set_margin_top(10);
    status_label.set_margin_bottom(10);

    STATUS_LABEL.with(|s| *s.borrow_mut() = Some(status_label.clone()));

    status_group.add(&WLabel(status_label));
    page.add(status_group.widget());

    page
}

/// Filtro de archivos tar (equivalente al Gtk.FileFilter del Python).
fn tar_filter() -> gtk::FileFilter {
    let filter = gtk::FileFilter::new();
    filter.set_name(Some("Archivo tar"));
    filter.add_pattern("*.tar");
    filter.add_pattern("*.tar.gz");
    filter.add_pattern("*.tar.zst");
    filter
}

/// Equivalente a _set_status.
fn set_status(msg: &str) {
    STATUS_LABEL.with(|s| {
        if let Some(label) = s.borrow().as_ref() {
            label.set_label(msg);
        }
    });
}

/// Dialogo de guardar backup (equivalente a _on_export).
fn on_export(btn: &gtk::Button) {
    let Some(window) = btn.root().and_downcast::<gtk::Window>() else {
        set_status("Error: no se pudo abrir el dialog (pagina sin root).");
        return;
    };

    let dialog = gtk::FileDialog::new();
    dialog.set_title("Guardar backup");

    let filter = tar_filter();
    let filters = gio::ListStore::new::<gtk::FileFilter>();
    filters.append(&filter);
    dialog.set_filters(Some(&filters));
    dialog.set_default_filter(Some(&filter));

    dialog.set_initial_name(Some("churros-backup.tar"));
    let home = std::env::var("HOME").unwrap_or_else(|_| "/root".to_string());
    dialog.set_initial_folder(Some(&gio::File::for_path(&home)));

    dialog.save(Some(&window), None::<&gio::Cancellable>, move |result| {
        let Ok(file) = result else {
            return;
        };
        let Some(path) = file.path().map(|p| p.to_string_lossy().to_string()) else {
            return;
        };

        set_status("Exportando...");

        match BackupService::export_to(&path) {
            Ok(_) => set_status(&format!("Backup guardado en {path}")),
            Err(e) => set_status(&format!("Error al exportar: {e}")),
        }
    });
}

/// Dialogo de abrir backup (equivalente a _on_import).
fn on_import(btn: &gtk::Button) {
    let Some(window) = btn.root().and_downcast::<gtk::Window>() else {
        set_status("Error: no se pudo abrir el dialog (pagina sin root).");
        return;
    };

    let dialog = gtk::FileDialog::new();
    dialog.set_title("Seleccionar backup de ChurrOS");

    let filter = tar_filter();
    let filters = gio::ListStore::new::<gtk::FileFilter>();
    filters.append(&filter);
    dialog.set_filters(Some(&filters));
    dialog.set_default_filter(Some(&filter));

    let home = std::env::var("HOME").unwrap_or_else(|_| "/root".to_string());
    dialog.set_initial_folder(Some(&gio::File::for_path(&home)));

    let btn = btn.clone();
    dialog.open(Some(&window), None::<&gio::Cancellable>, move |result| {
        let Ok(file) = result else {
            return;
        };
        let Some(path) = file.path() else {
            return;
        };
        let path = path.to_string_lossy().to_string();
        if !PathBuf::from(&path).is_file() {
            return;
        }
        confirm_import(&btn, &path);
    });
}

/// Confirmacion de importacion (equivalente a _confirm_import).
fn confirm_import(btn: &gtk::Button, path: &str) {
    let confirm = gtk::AlertDialog::builder()
        .message(
            "Esto reemplazara tu configuracion actual con la del archivo seleccionado. ¿Continuar?",
        )
        .modal(true)
        .buttons(["Cancelar", "Importar"])
        .build();

    let window = btn.root().and_downcast::<gtk::Window>();
    let path = path.to_string();

    confirm.choose(window.as_ref(), None::<&gio::Cancellable>, move |result| {
        let Ok(response) = result else {
            return;
        };
        if response != 1 {
            return;
        }

        set_status("Importando...");

        match BackupService::import_from(&path) {
            Ok(_) => set_status(
                "Configuracion importada. Reinicia las apps para ver todos los cambios.",
            ),
            Err(e) => set_status(&format!("Error al importar: {e}")),
        }
    });
}

/// Confirmacion de restablecimiento (equivalente a _on_reset).
fn on_reset(btn: &gtk::Button) {
    let dialog = gtk::AlertDialog::builder()
        .message(
            "Se borraran tus ajustes personales (tema, wallpaper, tipografia, dotfiles de niri/foot/fuzzel/mako/waybar) y se restauraran los defaults de ChurrOS. ¿Continuar?",
        )
        .modal(true)
        .buttons(["Cancelar", "Restablecer"])
        .build();

    let window = btn.root().and_downcast::<gtk::Window>();

    dialog.choose(window.as_ref(), None::<&gio::Cancellable>, move |result| {
        let Ok(response) = result else {
            return;
        };
        if response != 1 {
            return;
        }

        set_status("Restableciendo...");

        match BackupService::reset_to_defaults() {
            Ok(_) => set_status(
                "Configuraciones restablecidas a defaults. Reinicia las apps para ver todos los cambios.",
            ),
            Err(e) => set_status(&format!("Error al restablecer: {e}")),
        }
    });
}
